File: Backend/src/models/Order.ts
import {IOrder} from '../interfaces/IOrder';
import * as mongoose from 'mongoose';
import {OrderStatus} from "../enums/orderStatus";
import {employeeTypes} from "../enums/employeeTypes";


const { Schema } = mongoose;

const OrderSchema = new Schema({
    orderReferenceNo: {
        type: Number,
        unique: true
    },
    site: {
        type: Schema.Types.ObjectId,
        ref: 'site',
        required: true
    },
    supplier: {
        type: Schema.Types.ObjectId,
        ref: 'supplier',
        required: true
    },
    deliveryAddress:{
        type:String,
        required:true,
        trim:true
    },
    total:{
        type:Number,
        required:true
    },
    expectedDeliveryDate:{
        type:Date,
        required:true,
        trim:true
    },
    items: [
        {
            itemId:{
                type: Schema.Types.ObjectId,
                ref: 'items',
                required: true
            },
            qty:{
                type:Number,
                required:true
            },
            price:{
                type:Number,
                required:true
            },
            delivered: {
                type: Number,
                required: false,
                default: 0
            }
        }

    ],
    status: {
        type: String,
        enum: [OrderStatus.approved,OrderStatus.declined,OrderStatus.partiallyApproved,OrderStatus.pending,OrderStatus.placed,OrderStatus.referred,OrderStatus.returned,OrderStatus.waiting,OrderStatus.partiallyDelivered,OrderStatus.delivered],
        default: OrderStatus.pending
    },
    comments: [{
        type: Object
    }],
    orderNotes:{
        type:String,
        required:false,
        trim:true
    },
    createdBy: {
        type: Schema.Types.ObjectId,
        ref: 'users'
    },
    approvals: [
        {
            approvedBy:{
                type: Schema.Types.ObjectId,
                ref: 'users',
                required: false,
            },
            empType:{
                type: [String],
                enum: [employeeTypes.procurementStaff, employeeTypes.staffSupervisor, employeeTypes.siteManager],
                required: false
            },
            status:{
                type: String,
                enum: [OrderStatus.pending, OrderStatus.approved, OrderStatus.declined],
                required: false
            }
        }
    ],
}, { timestamps: true });

export default mongoose.model<IOrder & mongoose.Document>('orders', OrderSchema);

File: procuro_system/lib/models/Order.dart
class Order {
  static const ID = "_id";
  static const ORDERREFERENCENO = "orderReferenceNo";
  static const SITE = "site";
  static const SUPPLIER = "supplier";
  static const DELIVERYADDRESS = "deliveryAddress";
  static const EXPECTEDDELIVERYDATE = "expectedDeliveryDate";
  static const TOTAL = "total";
  static const ITEMS = "items";
  static const STATUS = "status";
  static const COMMENTS = "comments";

  final String id;
  final int orderReferenceNo;
  final Map<String, dynamic> site;
  final Map<String, dynamic> supplier;
  final String deliveryAddress;
  final String expectedDeliveryDate;
  final int total;
  final List<dynamic> items;
  final String status;
  final List<dynamic> comments;

  Order(
      {required this.id,
      required this.orderReferenceNo,
      required this.site,
      required this.supplier,
      required this.deliveryAddress,
      required this.expectedDeliveryDate,
      required this.total,
      required this.items,
      required this.status,
      required this.comments});

  factory Order.fromJson(Map<String, dynamic> json) {
    var order = Order(
      id: json[Order.ID],
      orderReferenceNo: json[Order.ORDERREFERENCENO],
      site: json[Order.SITE],
      supplier: json[Order.SUPPLIER],
      deliveryAddress: json[Order.DELIVERYADDRESS],
      expectedDeliveryDate: json[Order.EXPECTEDDELIVERYDATE],
      total: json[Order.TOTAL],
      items: json[Order.ITEMS],
      status: json[Order.STATUS],
      comments: json[Order.COMMENTS],
    );
    return order;
  }
}

class OrderList {
  final List<Order> orders;

  OrderList({required this.orders});

  factory OrderList.fromJson(List<dynamic> parsedJson) {
    List<Order> orders = <Order>[];
    orders = parsedJson.map((i) => Order.fromJson(i)).toList();
    return new OrderList(orders: orders);
  }
}
